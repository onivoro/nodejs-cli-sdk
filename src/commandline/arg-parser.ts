const DOUBLE_DASH = /^--/;

function isFlag(arg) {
  return arg && DOUBLE_DASH.test(arg);
}

function fromFlag(arg) {
  return arg && arg.replace(DOUBLE_DASH, '');
}

function helpRequested(args) {
  return args && args.length == 1 && args[0].replace(DOUBLE_DASH, '') === 'help';
}

export function argParser(args) {
  if (helpRequested(args)) {
    return null;
  }

  const output = { _: [] };
  let previous;
  let pushAll = false;

  for (let i = 0; i < args.length; i++) {
    const current = args[i];
    const next = args[i + 1];
    if (current === '--') {
      pushAll = true;
    }
    else if (pushAll) {
      output._.push(current);
    }
    else if (isFlag(current)) {
      const key = fromFlag(current);
      if (!next || isFlag(next)) {
        output[key] = true;
      } else {
        output[key] = next;
      }
    } else if (!previous || !isFlag(previous)) {
      output._.push(current)
    }

    previous = current;
  }

  return output;
}


