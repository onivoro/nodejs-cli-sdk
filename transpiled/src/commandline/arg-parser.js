"use strict";
exports.__esModule = true;
exports.argParser = void 0;
var DOUBLE_DASH = /^--/;
function isFlag(arg) {
    return arg && DOUBLE_DASH.test(arg);
}
function fromFlag(arg) {
    return arg && arg.replace(DOUBLE_DASH, '');
}
function helpRequested(args) {
    return args && args.length == 1 && args[0].replace(DOUBLE_DASH, '') === 'help';
}
function argParser(args) {
    if (helpRequested(args)) {
        return null;
    }
    var output = {};
    for (var i = 0; i < args.length; i++) {
        var current = args[i];
        var next = args[i + 1];
        if (isFlag(current)) {
            output[fromFlag(current)] = next;
        }
    }
    return output;
}
exports.argParser = argParser;
