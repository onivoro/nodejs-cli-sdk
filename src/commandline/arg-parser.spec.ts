import { argParser } from './arg-parser';

describe('argParser', () => {
  describe('when nothing provided', () => {
    it('returns empty object graph', () => {
      expect(argParser([])).toEqual({ _: [] });
    });
  });
  describe('when double-dash options/values provided', () => {
    it.each([
      [['--port', '1234']]
    ])('given %j', (args) => {
      expect(argParser(args)).toMatchSnapshot();
    });
  });
  describe('when double-dash options provided with no value', () => {
    it.each([
      [['--debug']],
      [['--debug', '--verbose']]
    ])('given %j', (args) => {
      expect(argParser(args)).toMatchSnapshot();
    });
  });
  describe('when args provided with no flags or options', () => {
    it.each([
      [['hot dogs', 'hamburgers', 'chicken']]
    ])('given %j', (args) => {
      expect(argParser(args)).toMatchSnapshot();
    });
  });

  describe('mixed cases', () => {
    it.each([      
      [['--find', 'caat', '--replace', 'cat', 'main.js']],
      [['--debug', '--verbose', '--', 'file1.txt']]
    ])('given %j', (args) => {
      expect(argParser(args)).toMatchSnapshot();
    });
  });
});