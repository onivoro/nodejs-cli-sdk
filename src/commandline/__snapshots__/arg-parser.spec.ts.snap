// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`argParser mixed cases given ["--debug","--verbose","--","file1.txt"] 1`] = `
Object {
  "_": Array [
    "file1.txt",
  ],
  "debug": true,
  "verbose": true,
}
`;

exports[`argParser mixed cases given ["--find","caat","--replace","cat","main.js"] 1`] = `
Object {
  "_": Array [
    "main.js",
  ],
  "find": "caat",
  "replace": "cat",
}
`;

exports[`argParser when args provided with no flags or options given ["hot dogs","hamburgers","chicken"] 1`] = `
Object {
  "_": Array [
    "hot dogs",
    "hamburgers",
    "chicken",
  ],
}
`;

exports[`argParser when double-dash options provided with no value given ["--debug","--verbose"] 1`] = `
Object {
  "_": Array [],
  "debug": true,
  "verbose": true,
}
`;

exports[`argParser when double-dash options provided with no value given ["--debug"] 1`] = `
Object {
  "_": Array [],
  "debug": true,
}
`;

exports[`argParser when double-dash options/values provided given ["--port","1234"] 1`] = `
Object {
  "_": Array [],
  "port": "1234",
}
`;
